hw/dv/sv/spi_agent/seq_lib/spi_device_extended_seq.sv

```
@@ -0,0 +1,137 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+// --------------------
+// Device sequence
+// This sequence acts as the device
+// with a DUT that acts as host.
+// the sequence will poll for read and write commands
+// for read commands it will sendback random data
+// --------------------
+
+
+class spi_device_extended_seq extends spi_device_seq;
+  `uvm_object_utils(spi_device_extended_seq)
+  `uvm_object_new
+
+  // read queue
+  spi_item rsp_q[$];
+  typedef enum {SPI_IDLE, SPI_ADDR, SPI_DATA } spi_fsm_e;
+  spi_fsm_e spi_state = SPI_IDLE;
+
+  virtual task body();
+
+    // wait for reset release
+    if (cfg.in_reset) wait (!cfg.in_reset);
+
+    fork
+      // get transactions from host
+      get_req();
+      // decode transactions
+      decode_cmd();
+      // transmit response
+      send_req(rsp_q);
+    join
+  endtask : body
+
+
+  virtual task decode_cmd();
+    spi_item item;
+    spi_item req = new();
+    spi_item req_clone;
+    spi_cmd_e  cmd;
+    bit [31:0] addr;
+    bit [31:0] data;
+    bit [31:0] addr_cnt = 0;
+    int        byte_cnt = 0;
+
+    forever begin
+      case (spi_state)
+        SPI_IDLE: begin
+          get_nxt_item(item);
+          // find start of transaction
+          if (item.first_byte) begin
+            // decode command
+            cmd = cmd_check(item.data.pop_front);
+            spi_state = SPI_ADDR;
+          end
```

Should there be an else clause where you would throw an error? That would indicate your FSM is out of whack and help debug issues (if they exist) quicker.

---

@sriyerg the current idea is just to ignore anything that is not a valid command - not to throw an error.
but If you feel it is needed I'd be happy to add a warning

---

My question is related `item.first_byte`. Is it acceptable for it to be 0 while you are in the `SPI_IDLE` state?

---

@sriyerg ahh I can see I add to the confusion with my answer - sorry about that.
it should throw a fatal if the command is invalid - but only if `item.first_byte == 1`
anything else is considered garbage and ignored.